lib/ddtrace/contrib/mongodb/parsers.rb

```diff
@@ -4,39 +4,56 @@ module Contrib
     module MongoDB
       module_function
 
-      # removes values for the given documents list
-      # NOTE: this is a rough estimation because it's possible to insert
-      # many values using different schemas; unfortunately to speed-up the
-      # parsing process this is the best guess.
-      # TODO: the normalization must be moved at Trace Agent level so that is
-      # faster and more accurate
-      def normalize_documents(documents)
-        return if documents.empty?
-
-        # always take the first element only to keep the resource cardinality
-        # and the normalization time low; the document is duplicated to avoid
-        # changing the Event query that is shared across the Monitoring system
-        document = documents.first.dup
-
-        # delete the unique identifier for this document
-        document.delete(:_id)
-        document.each do |key, _|
-          document[key] = '?'
+      # skipped keys are related to command names, since they are already
+      # extracted by the query_builder
+      SKIP_KEYS = [:_id].freeze
+      PLACEHOLDER = '?'.freeze
+
+      # returns a formatted and normalized query
+      def query_builder(command_name, database_name, command)
+        # always skip the command name
+        skip = SKIP_KEYS | [command_name.to_s]
+
+        result = {
+          operation: command_name,
+          database: database_name,
+          collection: command.values.first
+        }
+
+        command.each do |key, value|
+          result[key] = normalize_statement(value, skip) unless skip.include?(key)
         end
 
-        document
+        result
       end
 
-      # removes values from the given query keys
-      def normalize_query(frozen_query)
-        # the query is duplicated to avoid changing the Event query that is
-        # shared across the Monitoring system
-        query = frozen_query.dup
-        query.each do |key, _|
-          query[key] = '?'
+      # removes the values from the given query; this quantization recursively
+      # replace elements available in a given query, so that Arrays, Hashes and so
+      # on are compacted. It ensures a low cardinality so that it can be used
+      # as a Span resource.
+      # TODO: the quantization may not be enough, we still need to check the
+      # cardinality if it's reasonable
+      def normalize_statement(source, skip = [])
+        if source.is_a? Hash
+          obfuscated = {}
+          source.each do |key, value|
+            obfuscated[key] = normalize_value(value, skip) unless skip.include?(key)
+          end
```

Comment:
I'm totally fine with this change, makes sense, but at some point I think this is getting complicated enough that it's worth a small unit test (no need to overdo it, but a few examples showing it in action would help I think)

Reply:
sure! need to get some examples of output since the test is not done at this level but at the whole `Span` normalization.